linux: overlayfs support

freqlabs opened the pull request:
Backported https://github.com/openzfs/zfs/pull/9600 and https://github.com/openzfs/zfs/pull/12209 with new feature flags removed.

This means in cases where a pool is not cleanly exported you may end up
with loss of unflushed TX_RENAME_EXCHANGE and TX_RENAME_WHITEOUT
transactions and all transactions that follow them in the ZIL for that
dataset if the pool is then imported by a release build of ZFS that
does not recognize the new TX types.  A warning message is logged to
dmesg and the pool is fully usable, just missing the last changes.
Debug builds will panic when they fail to replay the full log.

The alternative is a pool that cannot be imported on any version of ZFS
lacking support for TX_RENAME_EXCHANGE and TX_RENAME_WHITEOUT replay
when uncleanly exported.  That prevents people from either upgrading
their pool for the ability to use overlayfs features or going back to a
previous boot environment after having upgraded the pool.

Jira: https://ixsystems.atlassian.net/browse/NAS-109036

ixhamza commented on module/os/linux/zfs/zfs_vnops_os.c:
```diff
@@ -2986,58 +3056,110 @@ zfs_rename(znode_t *sdzp, char *snm, znode_t *tdzp, char *tnm,
+		    (flags & FIGNORECASE ? TX_CI : 0), sdzp, sdl->dl_name,
+		    tdzp, tdl->dl_name, szp, wzp);
+		break;
+	default:
```
```rflags = 0``` for ```rename()``` system call will also fallback to default case. :thinking:

freqlabs replied:
I think that's the idea, `rename()` does `zfs_log_rename()`.

freqlabs commented on module/os/freebsd/zfs/zfs_vnops_os.c:
```diff
@@ -3473,14 +3473,17 @@ zfs_do_rename_impl(vnode_t *sdvp, vnode_t **svpp, struct componentname *scnp,
 	int error;
 	svp = tvp = NULL;
 
+	if (rflags != 0 || wo_vap != NULL)
```
`rflags` should always be zero and `wo_vap` always null on FreeBSD. We don't handle exchange or whiteout.

alek-p commented on AUTHORS:
```diff
@@ -19,6 +19,7 @@ CONTRIBUTORS:
     Albert Lee <trisk@nexenta.com>
     Alec Salazar <alec.j.salazar@gmail.com>
     Alejandro R. Sedeño <asedeno@mit.edu>
+    Aleksa Sarai <cyphar@cyphar.com>
```
this contributor should be on the next line

freqlabs replied:
I'll relay this to the upstream PR

freqlabs commented on module/zfs/zfs_replay.c:
```diff
@@ -663,13 +666,104 @@ zfs_replay_rename(void *arg1, void *arg2, boolean_t byteswap)
+	 * We should never reach this point because the feature is not
+	 * supported by non-Linux versions of OpenZFS.
+	 */
+	PANIC("TX_RENAME_EXCHANGE cannot be replayed on non-Linux systems.");
```
Should remove this panic.

freqlabs commented on module/zfs/zfs_replay.c:
```diff
@@ -663,13 +666,104 @@ zfs_replay_rename(void *arg1, void *arg2, boolean_t byteswap)
+	 * We should never reach this point because the feature is not
+	 * supported by non-Linux versions of OpenZFS.
+	 */
+	PANIC("TX_RENAME_WHITEOUT cannot be replayed on non-Linux systems.");
```
Should remove this panic.

freqlabs commented:
I reworked the backport of the commit adding the new TX types so they line up with the numbers in master (because we don't have TX_SETSAXATTR backported), and I adjusted the replay vectors to include an error function in the missing locations.

freqlabs commented:
Fixed zdb to handle the hole for setsaxattr, and dropped manpage changes for the features.
